### PR TITLE
Patterns: Add categories to user patterns, and allow filtering by these in site and post editor

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Add the ability to assign categories to user created patterns.

To make reviewing easier separate PRs will be created off this branch for:

- Filtering by user categories in site editor - https://github.com/WordPress/gutenberg/pull/53837
- Filtering by user categories in post editor inserter

## Why?
Currently users have no way of organising their synced and unsynced patterns.

## How?
Uses the [recently added](https://github.com/WordPress/gutenberg/pull/53163) `wp_patterns_category` custom taxonomy, and adds a taxonomy input box to the create pattern modal.

## Testing Instructions
- In the post editor add a block and then in the block toolbar overflow select the `Create pattern` option
- Add categories for the pattern in the create pattern model and add the pattern
- Go to `/wp-admin/edit.php?post_type=wp_block` and check that the categories are attached to the new pattern
- Try adding another pattern and try a mix of adding existing and new category names and make sure they are all saved
- Also try adding a new pattern from the site editor Patterns page

### Site editor testing instructions
 - Add several user patterns, synced and unsynced, with categories assigned (make sure to hit enter or add ',' after each category - this is the default tag input component behaviour - we will fix this in a follow up PR)
 - Go to the site editor patterns page and make sure the categories appear in the left nav panel
 - Make sure that when the categories are selected the correct patterns display
 - Make sure the theme patterns also still display as expected
 - Make sure that all the patterns display under 'All patterns' and that searching works as expected
 - Duplicate a theme pattern and make sure the new pattern is in the same category
 - Add a user pattern with category of `Footer` and make sure it displays along with the theme Footer patterns (be aware that although the TT3 theme footer patterns are listed under `Footers` the category slug is actually `footer` so the user category name needs to match this).

### Post Editor testing
- In the post editor add some synced and unsynced patterns with categories assigned
- Check that the patterns appear in the correct categories in the inserter patterns tab
- Check that the `All patterns` tab shows all patterns and that paging at the bottom of the tab works
- Check that the Source Filter select list works as expected and that correct patterns display for each selected filter
- Also check that when the `My patterns` source filter is selected that a `Sync type filter` appears at the top of the patterns list and works as expected
- Check that changing Source or Sync type filters when categories are selected resets the category to `All`
- Repeat all of the above in the patterns explorer modal
- With screenreader/voiceover running try selecting a pattern source from dropdown at top of patterns tab and check that the number of category buttons displayed is announced correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/d543a386-1b79-4496-84f9-7084642cdf2e

### CODE REVIEW

**Review by aaronrobertshaw — COMMENTED** [review]
Nice work here @glendaviesnz 👍 

This is looking good and tests as advertised for me.

✅ Pattern creation via the post editor works
✅ Creation via the site editor also works
✅ Pattern categories are correctly created or found, and assigned
✅ Post type admin page reflects the categories as they were assigned

Once we merge in the PRs built on this, we can give it a further holistic test before merging to trunk.

I'm looking forward to seeing all the pattern categorization in the wild!

**Comment by apeatling** [comment]
All worked great 👍 Part of me wanted to see the categories on the patterns view in the site editor.

✅ In the post editor add a block and then in the block toolbar overflow select the Create pattern option
✅ Add categories for the pattern in the create pattern model and add the pattern
✅ Go to /wp-admin/edit.php?post_type=wp_block and check that the categories are attached to the new pattern
✅ Try adding another pattern and try a mix of adding existing and new category names and make sure they are all saved
✅ Also try adding a new pattern from the site editor Patterns page


https://github.com/WordPress/gutenberg/assets/1464705/5dca4353-26ab-4db3-926c-3d274077d049

**Review by apeatling — APPROVED** [review]
All worked great 👍 Part of me wanted to see the categories on the patterns view in the site editor.

✅ In the post editor add a block and then in the block toolbar overflow select the Create pattern option
✅ Add categories for the pattern in the create pattern model and add the pattern
✅ Go to /wp-admin/edit.php?post_type=wp_block and check that the categories are attached to the new pattern
✅ Try adding another pattern and try a mix of adding existing and new category names and make sure they are all saved
✅ Also try adding a new pattern from the site editor Patterns page


https://github.com/WordPress/gutenberg/assets/1464705/5dca4353-26ab-4db3-926c-3d274077d049

**Review by aaronrobertshaw — COMMENTED** [review]
First things first, huge effort wrangling pattern categories @glendaviesnz 💪 

I've started testing the overall feature and it is coming together nicely! 

Unfortunately, I'm out of time today, so haven't yet gotten to thoroughly test all angles, or review the code completely, but thought I could share a couple of issues I've run into so far.

1. If you create a new pattern via the Patterns page's "Create pattern" menu item, then navigate back to the main Patterns page. You'll end up with the old `my-patterns` category selected even though that is no longer an option in the sidebar screen.
2. There's an inconsistent error thrown when searching patterns that crashes the page. 
    Screenshot 2023-09-14 at 8 32 40 pm
     https://github.com/WordPress/gutenberg/assets/60436221/dd6a4a9d-c5b7-474f-8c62-ae9d0741b4b4
3. In the post editor's pattern explorer modal, if you change the sync type filter there the modal closes unexpectedly.

https://github.com/WordPress/gutenberg/assets/60436221/663061ca-fbc4-4256-9b01-2f83a57011dd


Some other random thoughts were:
- It felt a bit odd to be changing the sync type filter in secondary panel in the inserter and then have the categories on the first panel change. It seemed like the sync type filter is visually scoped to the panel it is in.
- We should probably add pattern categories information to the pattern details screen when viewing a pattern
- Should we be enforcing that a pattern is named?
- Nit: There seems to be a 2px difference in height between the name input and the token input in the create pattern modal form.

I'll keep testing tomorrow and help out in any way I can.

It is looking pretty good though, I'm excited to see the feature land!

**Comment by glendaviesnz** [comment]
> In the post editor's pattern explorer modal, if you change the sync type filter there the modal closes unexpectedly.
> It felt a bit odd to be changing the sync type filter in secondary panel in the inserter and then have the categories on the first panel change. It seemed like the sync type filter is visually scoped to the panel it is in.

These two were related. The idea with updating the categories was that currently the default was not to show categories that were not populated. However I agree the resulting UX is a little odd, and in fact it probably works better to just show 0 results for the cases where a sync filter change means the category is not populated. I have a [PR here](https://github.com/WordPress/gutenberg/pull/54480) that takes this approach, and it also fixes the issues with the modal randomly closing which was caused by a rerender triggered by the category updates.

**Comment by glendaviesnz** [comment]
> If you create a new pattern via the Patterns page's "Create pattern" menu item, then navigate back to the main Patterns page. You'll end up with the old my-patterns category selected even though that is no longer an option in the sidebar screen.

Fixed [here](https://github.com/WordPress/gutenberg/pull/53835/commits/2d36539a501d4c6885e1896eef67d2b5ecb8f633).

**Comment by glendaviesnz** [comment]
> Nit: There seems to be a 2px difference in height between the name input and the token input in the create pattern modal form.

This was just the default sizing of both of these from the components package. Fixed [here](https://github.com/WordPress/gutenberg/pull/53835/commits/033974b74556e52d8c9284d400e3b8f13e38c828).

**Comment by glendaviesnz** [comment]
> There's an inconsistent error thrown when searching patterns that crashes the page.

Fixed [here](https://github.com/WordPress/gutenberg/pull/53835/commits/7dbcdd1e0df9c073b2ab7456e4a42f073c906cd1). @kevin940726 with your experience with the iFrame events can you see any issues with this fix?

**Comment by glendaviesnz** [comment]
> We should probably add pattern categories information to the pattern details screen when viewing a pattern
> Should we be enforcing that a pattern is named?

Agree, but we should do these in follow-up PRs as not critical.

**Comment by aaronrobertshaw** [comment]
The latest fixes are working pretty well.

> 1. my-patterns category incorrectly set when navigating back from newly created pattern

✅ Confirmed that the `all-patterns` category id is set when navigating back in this instance

❓ This did highlight though we have a regression where navigating back from viewing a pattern loses the selected category in the sidebar screen. Confirmed it works on trunk but not this branch.


> 2. Crash when searching patterns

I can't replicate the issue now after the fix which LGTM but wise to wait for @kevin940726's expert thoughts.

> 3. Unexpected closure of the pattern explorer

The changes in https://github.com/WordPress/gutenberg/pull/54480 test pretty well for me however I did hit a small issue there and comment on that in the review. Overall though I think the change in UX is for the better.

With all that said, the pattern categories feature is getting pretty close. Once we merge in https://github.com/WordPress/gutenberg/pull/54480 we might just need a few extra sets of eyes to land it.

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/use-pattern-details.js [hunk 1]
```diff
@@ -82,6 +88,18 @@ export default function usePatternDetails( postType, postId ) {
...
+				.filter( ( category ) => patternCategories.get( category ) )
+				.map( ( category ) => patternCategories.get( category ).label );
+			details.push( {
+				label: __( 'Categories' ),
+				value:
+					categories.length > 0
+						? categories.join( ', ' )
+						: __( 'Uncategorized' ),
+			} );
+		}
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/use-pattern-details.js]: With the initial check for `.length > 0` we'll not reach the `Uncategorized` label here. 

I think we should be adding whether a pattern is currently uncategorized to the details screen so best if we can fix this up.

**Comment by kevin940726** [comment]
> > There's an inconsistent error thrown when searching patterns that crashes the page.
> 
> Fixed [here](https://github.com/WordPress/gutenberg/pull/53835/commits/7dbcdd1e0df9c073b2ab7456e4a42f073c906cd1). @kevin940726 with your experience with the iFrame events can you see any issues with this fix?

I think the root cause is that when an iframe gets removed it still runs the render function with the stale `iframeDocument` for some reason, but we should also clean that up to prevent memory leak. This is a much bigger bug fix though and we'll need time to find the best way to solve this correctly. In the meantime, I think a quick fix of checking `defaultView` in this PR is good and we can at least prevent the page crash 👍.

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/use-pattern-details.js (hunk 1 above)
Comment (glendaviesnz) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/use-pattern-details.js]: Ah, with that `uncategorized` I was thinking of instances where the category id may be assigned but may  not match the existing categories, eg. if user assigned a category, but then deleted it from category management page, but it looks like WP nicely tidies that up, so I have [updated to correctly assign uncategorized only if there are no category ids](https://github.com/WordPress/gutenberg/pull/53835/commits/11ae9f6c319dcf1aa041e12d0df6548d81653a8d).

**Comment by glendaviesnz** [comment]
> We should probably add pattern categories information to the pattern details screen when viewing a pattern

have [added this](https://github.com/WordPress/gutenberg/pull/53835/commits/f4de28af2f010c2b168cf8b935e969e6ff2190b3).

**Comment by glendaviesnz** [comment]
>  This did highlight though we have a regression where navigating back from viewing a pattern loses the selected category in the sidebar screen. Confirmed it works on trunk but not this branch.

This should be [fixed now.](https://github.com/WordPress/gutenberg/pull/53835/commits/11cbc987451ce7422a8871076f691ab9733de8ac)

**Review by aaronrobertshaw — COMMENTED** [review]
The latest fixes to indicating category selection and uncategorized detail are working well for me.

Once https://github.com/WordPress/gutenberg/pull/54480 is merged in I'll give this another smoke test and we should be ready to merge.

Edit: Looks like it merged while I was typing. I'll pull the latest and give it another run.

**Review by aaronrobertshaw — APPROVED** [review]
I've re-tested following the PR test instructions and didn't run into any problems. It's looking pretty good, nice work!

Other than a rebase to resolve the conflict, is there anything else you think needs to be addressed before we land this one? 

I'd like to think any issues should be addressable in follow-ups. What do you think?

**Comment by glendaviesnz** [comment]
I have rebased. There is now an error in console on site editor categories page:
Screenshot 2023-09-15 at 4 54 21 PM
but have checked and this also appears on trunk.

If all the tests pass I think we should be good to merge, and as you note we can sort any additional issues that might arise in follow ups.

**Comment by glendaviesnz** [comment]
Looks like the console error might have been introduced by https://github.com/WordPress/gutenberg/pull/54450

**Comment by aaronrobertshaw** [comment]
> but have checked and this also appears on trunk

Confirmed. This doesn't appear to have been caused during the conflict resolution but introduced via https://github.com/WordPress/gutenberg/pull/54450. 

A separate fix for that issue is probably better for historical purposes. So I don't think it is a blocker for this PR.

**Comment by ramonjd** [comment]
Not a big deal, just adding it here coz it fresh. 

When trying to add a category that already exists, the API returns a 400 Bad Request error:


https://github.com/WordPress/gutenberg/assets/6458278/e74eaec8-f4cf-446c-a0d6-a4c1b4e609aa

If this is a problem, maybe we could fetch the taxonomy records and check if it exists before adding it, similar to how post taxonomies do it?

https://github.com/WordPress/gutenberg/blob/900439e2d123eff21525c1bcc5e1349c787e4e0f/packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js#L277-L277

**Comment by glendaviesnz** [comment]
@ramonjd that approach was copied from the [post taxonomies flat-term-selector ](https://github.com/WordPress/gutenberg/blob/900439e2d123eff21525c1bcc5e1349c787e4e0f/packages/editor/src/components/post-taxonomies/flat-term-selector.js#L148), so I assumed it was a legit approach, but happy to review this approach if you have concerns about it.

**Comment by ramonjd** [comment]
>  I assumed it was a legit approach, but happy to review this approach if you have concerns about it.

No worries, I was just flagging it. Thanks for the explainer. 👍🏻

**Comment by WunderBart** [comment]
👋 This PR seems to have caused a large increase (58%) in the editor's typing performance metric:

Screenshot 2023-09-18 at 12 16 53

_From https://www.codevitals.run/project/gutenberg._

cc @youknowriad

File: packages/block-editor/src/components/inserter/hooks/use-patterns-state.js [hunk 2]
```diff
@@ -68,7 +80,7 @@ const usePatternsState = ( onInsert, rootClientId ) => {
 		[ createSuccessNotice, onInsert ]
 	);
 
-	return [ patterns, allCategories, onClickPattern ];
+	return { patterns, allCategories, onClickPattern };
```
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/hooks/use-patterns-state.js:83]: Was this change necessary at all? Now it's not consistent anymore with useBlocksState.

**Review by youknowriad — COMMENTED** [review]
Not sure exactly what caused the regression but it's meaningful indeed. Worth looking into.

Also the PR is a bit too big. I wonder if it could have shipped in smaller bits.

File: packages/core-data/src/selectors.ts [hunk 3]
```diff
@@ -1222,6 +1236,28 @@ export function getBlockPatternCategories( state: State ): Array< any > {
...
+	const patternCategoriesMap = new Map< number, UserPatternCategory >();
+	state.userPatternCategories?.forEach(
+		( userCategory: UserPatternCategory ) =>
+			patternCategoriesMap.set( userCategory.id, userCategory )
+	);
+	return {
+		patternCategories: state.userPatternCategories,
+		patternCategoriesMap,
+	};
+}
```
Comment (Mamaduka) [inline packages/core-data/src/selectors.ts:1259]: This needs to be memorized. The `useSelect` noticed from this screenshot is caused by this selector - https://github.com/WordPress/gutenberg/pull/53835#issuecomment-1720669088.

Aside from this, is there a reason we're creating a map based on the state inside a selector? The task is better suited for consumers, in my opinion.

File: packages/block-editor/src/components/inserter/hooks/use-patterns-state.js (hunk 2 above)
Comment (glendaviesnz) [reply packages/block-editor/src/components/inserter/hooks/use-patterns-state.js:83]: The object return makes for a nicer API, eg. avoids the likes of `const [ patterns, , onClickPattern ] = `. An array return makes sense for hooks like useState where you are always going to use both values and want to easily give them context-specific names, but not so much in instances like this hook. 

I don't have a strong opinion on this though. If keeping it consistent with `useBlocksState` is thought to be more important then there is a [revert PR here](https://github.com/WordPress/gutenberg/pull/54582).

File: packages/core-data/src/selectors.ts (hunk 3 above)
Comment (glendaviesnz) [reply packages/core-data/src/selectors.ts:1259]: It doesn't look like this selector was causing that screenshot error as that error was there on trunk prior to this merge, and still there after fixing this selector - we will look at that separately. 

This selector was causing the useBlockEditorSettings useMemo return value to be rerun with each call though, which happens on each keypress, so have [moved the map into the consumers as suggested](https://github.com/WordPress/gutenberg/pull/54576). While investigating this I also discovered another change which was causing this useMemo to be reevaluated with every call of the hook. This has been [fixed here](https://github.com/WordPress/gutenberg/pull/54580).

**Comment by glendaviesnz** [comment]
https://github.com/WordPress/gutenberg/pull/54580 seems to have resolved the typing issue:

Screenshot 2023-09-19 at 4 18 52 PM

thanks @WunderBart for highlighting this, and @Mamaduka for the pointer to that selector.

**Comment by Mamaduka** [comment]
Thanks for the follow-ups, @glendaviesnz!

File: packages/core-data/src/selectors.ts (hunk 3 above)
Comment (Mamaduka) [reply packages/core-data/src/selectors.ts:1259]: I meant the warning, visible at the top of the error and pointing to the `useBlockEditorSettings`.

## Screenshot
CleanShot 2023-09-19 at 09 46 31
